src/java.base/share/classes/java/io/InputStream.java

```diff
@@ -542,7 +545,11 @@ public long skip(long n) throws IOException {
         }
 
         int size = (int)Math.min(MAX_SKIP_BUFFER_SIZE, remaining);
-        byte[] skipBuffer = new byte[size];
+
+        if ((skipBuffer == null) || (skipBuffer.length < size)) {
```

Comment:
Isn't it the case that the length of the global `skipBuffer` is non-decreasing? Thus skipping 6 bytes after skipping 10 will not result in a new buffer allocation.

Even so, it does appear that prior buffer allocations could be "lost" due to concurrent use of `skip` resulting in a "minor" memory leak in the heap.

Reply:
> Even so, it does appear that prior buffer allocations could be "lost" due to concurrent use of skip resulting in a "minor" memory leak in the heap.

The minor leak, though maybe somewhat wasteful, is the necessary cost for thread safety as opposed to locking or synchronizing to reliably retrieve a previously allocated array from another thread.

Another concern about this change is that now the skip buffer is no longer immediately GC'd when exiting the scope but will live on till the input stream is no longer reachable. For instance, if an input stream is skipped and then kept for long, this may be a performance regression instead for the increased memory use.

Reply:
Is this change proposed as the result of some experience with particular apps and streams or just observation of the implementation?

Is there any information about what kind of underlying streams do not support skip directly
and what is the distribution of sizes expected to be 'skipped'?
GIven the object instance overhead of an array, I'd be inclined to suggest a minimum size that would
cover all of the 'small' skipping that could occur. And if that's not big enough, jump to the MAX_SKIP_BUFFER_SIZE. To keep the code simple, I'd start at 128bytes and jump to the max if that's not big enough.  There is precious little actual information available to fine tune.

Reply:
> Is this change proposed as the result of some experience with particular apps and streams or just observation of the implementation?

just observation of the implementation?

> Is there any information about what kind of underlying streams do not support skip directly and what is the distribution of sizes expected to be 'skipped'? GIven the object instance overhead of an array, I'd be inclined to suggest a minimum size that would cover all of the 'small' skipping that could occur. And if that's not big enough, jump to the MAX_SKIP_BUFFER_SIZE. To keep the code simple, I'd start at 128bytes and jump to the max if that's not big enough. There is precious little actual information available to fine tune.

It whould make the problem @liach said above heavier.

For example, if people just invoke 129, then the imput stream object hold a 8192 length array until it die.

That sounds much too wasting.

Reply:
Without specific information about use cases, there isn't enough information to
craft a good algorithm/solution and simplicity is preferred. 
The MAX_SKIP_BUFFER_SIZE is 2048 (not 8192).

What subclasses of InputStream in the JDK do not override skip(n)?
Most sequential streams are open for a relatively short period of time, the lifetime of the 
memory for the buffer won't change the memory usage enough to notice.

If the concern is about tying up memory then allocate the buffer once and
don't resize it.  Each resize consumes extra memory and gc cycles to reclaim the last buffer.
Use the requested size but at least nnn and at most MAX_SKIP_BUFFER_SIZE.

Reply:
> Without specific information about use cases, there isn't enough information to craft a good algorithm/solution and simplicity is preferred. The MAX_SKIP_BUFFER_SIZE is 2048 (not 8192).
> 
> What subclasses of InputStream in the JDK do not override skip(n)? Most sequential streams are open for a relatively short period of time, the lifetime of the memory for the buffer won't change the memory usage enough to notice.
> 
> If the concern is about tying up memory then allocate the buffer once and don't resize it. Each resize consumes extra memory and gc cycles to reclaim the last buffer. Use the requested size but at least nnn and at most MAX_SKIP_BUFFER_SIZE.

@RogerRiggs Sounds reasonable and applied. Should we change the implementation in Reader class as well?

Reply:
> Shouldn't be too problematic, as most skip usages in JDK as I see are skipping small number of bytes like 2 or 4, or like skipping over attributes of Java class files. A minimum skip buffer size isn't that helpful, as I don't think we often see skip calls with slowly incremental sizes.

Yep usually people skip for fixed length... It is rare to see people skip larger and larger...